File: src/main/java/org/stt/Streams.java
package org.stt;

import java.util.HashSet;
import java.util.Set;
import java.util.function.Function;
import java.util.function.Predicate;

public class Streams {
    private Streams() {
    }

    public static <T> Predicate<T> distinctByKey(Function<T, Object> keyExtractor) {
        Set<Object> visited = new HashSet<>();
        return item -> visited.add(keyExtractor.apply(item));
    }
}

File: src/main/kotlin/org/stt/Streams.kt
package org.stt

import java.util.*
import java.util.function.Predicate

object Streams {

    fun <T> distinctByKey(keyExtractor: (T) -> Any): Predicate<T> {
        val visited = HashSet<Any>()
        return Predicate { item -> visited.add(keyExtractor(item)) }
    }
}
